src/ripple/overlay/impl/OverlayImpl.cpp

```diff
@@ -1298,6 +1324,65 @@ OverlayImpl::relay(
     return {};
 }
 
+void
+OverlayImpl::relay(
+    uint256 const& hash,
+    protocol::TMTransaction& m,
+    std::set<Peer::id_t> const& toSkip)
+{
+    auto const sm = std::make_shared<Message>(m, protocol::mtTRANSACTION);
+
+    if (app_.config().TX_REDUCE_RELAY_ENABLE)
+    {
+        // returns active peers, which are not in toSkip, and
+        // a number of peers with not enabled tx reduce-relay
+        // feature and which are not in toSkip.
+        auto [peers, notEnabled] = getActivePeers(toSkip);
+
+        // select a fraction of active peers for relaying
+        auto toSelect = static_cast<uint32_t>(
+            app_.config().TX_RELAY_TO_PEERS * peers.size() / 100);
+
+        txMetrics_.addMetrics(toSelect, toSkip.size(), notEnabled);
+
+        // exclude from peers, which should be randomly selected for relaying,
+        // peers in toSkip and peers with not enabled tx reduce-relay feature
+        auto const exclude = toSkip.size() + notEnabled;
+        toSelect = (toSelect > exclude) ? (toSelect - exclude) : 0;
```

Comment:
Could you please explain the calculation of `toSelect`? I think the part that confused me is the definition of `TX_RELAY_TO_PEERS`, i.e. "Percentage of peers to select from total peers" as in config.h. My understanding is that if a node has 100 peers and `TX_RELAY_TO_PEERS`=50, then the node wants to make sure 50 of its peers have the Tx. Either they have the Tx already or the node will forward the Tx to them. 

Now consider a case that `toSkip.size()` = 40 and `notEnabled` = 0. Then `exclude` = 40, `peers.size()` = 60, and `TX_RELAY_TO_PEERS * peers.size() / 100` = 50*60/100 = 30. And 30 < 40, so `toSelect` will be 0. So no forwarding now. Hope my calculation is correct.